New support for the STM  NUCLEO_L433RC-P board

helmut64 opened the pull request:
Dear ST and MBED team, 
I added support for a new the target NUCLEO_L433RC.
The L433 is pretty much like the L432,  but STDIO pins and peripherals are different.
I tested this using the official NUCLEO_L433RC board from STM.
I have not done the job for export_test.py and tests.py as documented in “steps-to-create-a-new-STM32-platform” because the documentation does not match file files.
Regards Helmut

bcostm commented:
Thanks a lot !  We'll review the code and run all the tests.

bcostm commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L433xC/TARGET_NUCLEO_L433RC/PinNames.h:
```diff
@@ -0,0 +1,199 @@
+  PB_14 = 0x1E,
+  PB_15 = 0x1F,
+
+#ifndef STM32L433_48PINS // 48 pin versions don't have PC0-PC15 pins
```
Where is this define set ?

helmut64 replied:
The define STM32L433_48PINS can be set by users to limited to pins to the 48 pin versions of the STM433 MCUs, by default this is included because the Nucleo board uses the 64 pin version.
The define STM32L433_100PINS can be set by users to extend the pins to the 100 pin versions of the STM433 MCUs.

The is reasonable to be more flexible for the pin definition which allows many more 433 MCUs to be used.

The __weak of PinMaps in the PeripheralPins.c allows to extend this for other MCUs, another option would be to use the PeripheralPins.c for the 100 pin MCU version which is a superset of the 64 pin version.

bcostm replied:
OK. The __weak is a good idea and  I prefer this solution.

bcostm commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L433xC/TARGET_NUCLEO_L433RC/PeripheralPins.c:
```diff
@@ -0,0 +1,251 @@
+    {PC_1,       ADC_1,    STM_PIN_DATA_EXT(STM_MODE_ANALOG, GPIO_NOPULL, 0, 2, 0)}, // ADC1_IN2
+    {PC_2,       ADC_1,    STM_PIN_DATA_EXT(STM_MODE_ANALOG, GPIO_NOPULL, 0, 3, 0)}, // ADC1_IN3
+    {PC_3,       ADC_1,    STM_PIN_DATA_EXT(STM_MODE_ANALOG, GPIO_NOPULL, 0, 4, 0)}, // ADC1_IN4
+    {PC_4,       ADC_1,    STM_PIN_DATA_EXT(STM_MODE_ANALOG, GPIO_NOPULL, 0, 13, 0)}, // ADC1_IN13
```
Please add also:
{PC_5,       ADC_1,    STM_PIN_DATA_EXT(STM_MODE_ANALOG, GPIO_NOPULL, 0, 14, 0)}, // ADC1_IN14

The file I gave you was for the NUCLEO_L433RC_P

bcostm commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L433xC/TARGET_NUCLEO_L433RC/PeripheralPins.c:
```diff
@@ -0,0 +1,251 @@
+    {PB_10,      UART_3,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART3)},
+    {PB_11,      LPUART_1,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF8_LPUART1)},
+    {PC_1,       LPUART_1,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF8_LPUART1)},
+    {PC_4,       UART_3,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART3)},
```
Please add:
`{PC_5,       UART_3,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART3)},`

bcostm commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L433xC/TARGET_NUCLEO_L433RC/PeripheralPins.c:
```diff
@@ -0,0 +1,251 @@
+    {PB_7,       UART_1,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART1)},
+    {PB_10,      LPUART_1,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF8_LPUART1)},
+    {PB_11,      UART_3,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART3)},
+    {PC_0,       LPUART_1,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF8_LPUART1)},
```
Please add:
`{PC_5,       UART_3,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART3)},`

bcostm commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L433xC/TARGET_NUCLEO_L433RC/PeripheralPins.c:
```diff
@@ -0,0 +1,251 @@
+    {PB_1,       UART_3,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART3)},
+    {PB_3,       UART_1,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART1)},
+    {PB_12,      LPUART_1,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF8_LPUART1)},
+    {PB_14,      UART_3,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART3)},
```
Please add:
`{PD_2,       UART_3,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART3)},`

bcostm commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L433xC/TARGET_NUCLEO_L433RC/PeripheralPins.c:
```diff
@@ -0,0 +1,251 @@
+
+//*** PWM ***
+// TODO: Timers/PWM for the L433xx need to be reviewed/completed,
+//       Check for TIM6/TIM7 of the 433
```
I don't see any PWM pins on TIM6/TIM7 in the DataSheet.

Can you add the "_ALT0" on the pins that we don't use ? (your idea :smile:)

bcostm commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L433xC/TARGET_NUCLEO_L433RC/PeripheralPins.c:
```diff
@@ -0,0 +1,251 @@
+//*** SERIAL ***
+
+__weak const PinMap PinMap_UART_TX[] = {
+//     {PA_2,       LPUART_1,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF8_LPUART1)},
```
Can you add a comment:
`// Connected to STDIO_UART_TX`

bcostm commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L433xC/TARGET_NUCLEO_L433RC/PeripheralPins.c:
```diff
@@ -0,0 +1,251 @@
+};
+
+__weak const PinMap PinMap_UART_RX[] = {
+//    {PA_3,       LPUART_1,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF8_LPUART1)},
```
Can you add a comment:
`// Connected to STDIO_UART_RX`

bcostm commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L433xC/TARGET_NUCLEO_L433RC/PeripheralPins.c:
```diff
@@ -0,0 +1,251 @@
+    {PA_1,       UART_2,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART2)},
+    {PA_12,      UART_1,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART1)},
+    {PA_15,      UART_3,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART3)},
+    {PB_1,       LPUART_1,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF8_LPUART1)},
```
Can you add the "_ALTx" on the pins that we don't use ?

bcostm commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L433xC/TARGET_NUCLEO_L433RC/PeripheralPins.c:
```diff
@@ -0,0 +1,251 @@
+
+__weak const PinMap PinMap_UART_CTS[] = {
+    {PA_0,       UART_2,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART2)},
+    {PA_6,       LPUART_1,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF8_LPUART1)},
```
Can you add the "_ALTx" on the pins that we don't use ?

bcostm commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L433xC/TARGET_NUCLEO_L433RC/PeripheralPins.c:
```diff
@@ -0,0 +1,251 @@
+    {PA_6,       UART_3,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART3)},
+    {PA_11,      UART_1,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART1)},
+    {PB_4,       UART_1,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART1)},
+    {PB_13,      LPUART_1,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF8_LPUART1)},
```
Can you add the "_ALTx" on the pins that we don't use ?

bcostm commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L433xC/TARGET_NUCLEO_L433RC/PeripheralPins.c:
```diff
@@ -0,0 +1,251 @@
+    {PA_7,       SPI_1,    STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF5_SPI1)},
+    {PA_12,      SPI_1,    STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF5_SPI1)},
+    {PB_5,       SPI_1,    STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF5_SPI1)},
+    {PB_5,       SPI_3,    STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF6_SPI3)},
```
Can you add the "_ALTx" on the pins that we don't use ?

bcostm commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L433xC/TARGET_NUCLEO_L433RC/PeripheralPins.c:
```diff
@@ -0,0 +1,251 @@
+__weak const PinMap PinMap_SPI_SSEL[] = {
+    {PA_4,       SPI_1,    STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF5_SPI1)},
+    {PA_4,       SPI_3,    STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF6_SPI3)},
+    {PA_15,      SPI_1,    STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF5_SPI1)},
```
Can you add the "_ALTx" on the pins that we don't use ?

bcostm commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L433xC/TARGET_NUCLEO_L433RC/PeripheralPins.c:
```diff
@@ -0,0 +1,251 @@
+__weak const PinMap PinMap_SPI_SCLK[] = {
+    {PA_1,       SPI_1,    STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF5_SPI1)},
+    {PA_5,       SPI_1,    STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF5_SPI1)},
+    {PB_3,       SPI_1,    STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF5_SPI1)},
```
Can you add the "_ALTx" on the pins that we don't use ?

bcostm commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L433xC/TARGET_NUCLEO_L433RC/PeripheralPins.c:
```diff
@@ -0,0 +1,251 @@
+__weak const PinMap PinMap_SPI_MISO[] = {
+    {PA_6,       SPI_1,    STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF5_SPI1)},
+    {PA_11,      SPI_1,    STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF5_SPI1)},
+    {PB_4,       SPI_1,    STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF5_SPI1)},
```
Can you add the "_ALTx" on the pins that we don't use ?

bcostm commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L433xC/TARGET_NUCLEO_L433RC/PeripheralPins.c:
```diff
@@ -0,0 +1,251 @@
+ * I declared all mapping tables as weak to allow custom overwrites for similar MCU models
+ * The busy GPIOs e.g. PA_3/3 are still declared to allow reusing them for other IO.
+ */
+
```
Can you please replace the 2 above notes with this one:
```
//==============================================================================
// Notes
//
// - The pins mentionned Px_y_ALTz are alternative possibilities which use other
//   HW peripheral instances. You can use them the same way as any other "normal"
//   pin (i.e. PwmOut pwm(PA_7_ALT0);). These pins are not displayed on the board
//   pinout image on mbed.org.
//
// - The pins which are connected to other components present on the board have
//   the comment "Connected to xxx". The pin function may not work properly in this
//   case. These pins may not be displayed on the board pinout image on mbed.org.
//   Please read the board reference manual and schematic for more information.
//
// - All mapping tables are declared as weak to allow custom overwrites for similar MCU models.
//
//==============================================================================
```

bcostm commented on targets/targets.json:
```diff
@@ -1345,6 +1345,23 @@
+                "macro_name": "CLOCK_SOURCE"
+            }
+        },
+        "detect_code": ["0770"],
```
I have the ID `0779 `when I connect my NUCLEO-L433RC-P.

Can you confirm the ID you see in the MBED.HTM file (first 4 digits in the code field) ?

`<meta http-equiv="refresh" content="0; url=http://mbed.org/device/?code=07790221172C637A2868F53E"/> `

helmut64 replied:
Yes my Code is: code=07790221172F6363286BF527

bcostm commented:
As the board is a NUCLEO_L433RC-P can you please change the name is this PR title and also in the files ?

bcostm commented on targets/targets.json:
```diff
@@ -1345,6 +1345,23 @@
         "release_versions": ["2", "5"],
         "device_name": "STM32L432KC"
     },
+    "NUCLEO_L433RC": {
```
NUCLEO_L433RC_P

bcostm commented on tools/build_travis.py:
```diff
@@ -54,6 +54,7 @@
     { "target": "NUCLEO_F411RE",     "toolchains": "GCC_ARM", "libs": ["dsp", "usb"] },
     { "target": "NUCLEO_F412ZG",     "toolchains": "GCC_ARM", "libs": ["dsp"] },
     { "target": "NUCLEO_L432KC",     "toolchains": "GCC_ARM", "libs": ["dsp"] },
+    { "target": "NUCLEO_L433RC",     "toolchains": "GCC_ARM", "libs": ["dsp"] },
```
NUCLEO_L433RC_P

bcostm commented on tools/export/sw4stm32/__init__.py:
```diff
@@ -244,6 +244,11 @@ class Sw4STM32(GNUARMEclipse):
             'name': 'NUCLEO-L432KC',
             'mcuId': 'STM32L432KCUx'
         },
+        'NUCLEO_L433RC':
```
NUCLEO_L433RC_P

bcostm commented on tools/export/sw4stm32/__init__.py:
```diff
@@ -244,6 +244,11 @@ class Sw4STM32(GNUARMEclipse):
         },
+        'NUCLEO_L433RC':
+        {
+            'name': 'NUCLEO-L433RC',
```
NUCLEO-L433RC-P

bcostm commented on tools/export/sw4stm32/__init__.py:
```diff
@@ -244,6 +244,11 @@ class Sw4STM32(GNUARMEclipse):
+        'NUCLEO_L433RC':
+        {
+            'name': 'NUCLEO-L433RC',
+            'mcuId': 'STM32L433RCUx'
```
STM32L433RCTxP

helmut64 commented:
@bcostm I did all the changes you proposed, however no ALTx changes.
I don't get the ALTx concept, maybe you can explain it to me.
Thank you for reviewing the changes and getting this done together.

ohagendorf commented:
@helmut64 maybe my description in the introductory PR #3181 helps

helmut64 commented:
@ohagendorf Danke, now I understand, the PinMap table as an unique index for each pin therfore we use alternate names for multiple pin options. Working on it.

bcostm commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L433xC/TARGET_NUCLEO_L433RC/PeripheralPins.c:
```diff
@@ -0,0 +1,251 @@
+
+__weak const PinMap PinMap_UART_RX[] = {
+//    {PA_3,       LPUART_1,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF8_LPUART1)},
+    {PA_3,       UART_2,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF7_USART2)},
```
Can you add a comment:
`// Connected to STDIO_UART_RX`

bcostm commented:
Thanks @helmut64 
Concerning the pins PC_4, PC_5 and PD_2, you have to remove them because they are not presetn on the NUCLEO_L433RC_P. Your original file version was correct. I thought you were working on the NUCLEO_L433RC (without the '-P'). Sorry about that.

helmut64 commented:
@bcostm I keep the PC_5, PD_2 because I use the STM433RCT for my projects. This is not hurting and helps me.
I added the _ALT mappings, this is a good solution.
All your feedback has been incorporated.

helmut64 commented:
@bcostm I reviewed the targets.json file for our new target, here are some questions:
**a) "bootloader_supported": true**
This is missing, it is included in the NUCLEO_L476RG/NUCLEO_L486RG targets, I don't see any reason not to enable it as well for the 433RC_P board. Any ideas?

*b) SERIAL_ASYNCH*
Again there is any reason this should be disabled, because the 433 uses the same serial as the other L4 MCUs. Maybe there is a difference in the DMA channels.

**c)  "release_versions": ["2", "5"],**
Is this correct, I just copied it from the 432KC as a template.

Your advise will be appreciated.

LMESTM commented:
> a)  "bootloader_supported": true
This is missing, it is included in the NUCLEO_L476RG/NUCLEO_L486RG targets, I don't see any reason not to enable it as well for the 433RC_P board. Any ideas?

Hi - the bootloader supports requires few adaptation of the memory maps, like in #4474 / #4469 (former #4063) . That's a good idea to consider those changes from the start indeed.

jeromecoutant commented:
> c) "release_versions": ["2", "5"],

Yes, this is correct.

bcostm commented:
> b) SERIAL_ASYNCH
> Again there is any reason this should be disabled, because the 433 uses the same serial as the other L4 MCUs. Maybe there is a difference in the DMA channels.

We have discussed internally and the SERIAL_ASYNCH should be ok on this target. It is not related to the DMA as it is not used. So you can add it in the targets.json.

bcostm commented:
mbedOS 2 tests are OK
```
+--------+-----------------+-----------+-------------+---------------------------------------+--------------+---------+
| Result | Target          | Toolchain | Test ID     | Test Description                      | Elapsed Time | Timeout |
+--------+-----------------+-----------+-------------+---------------------------------------+--------------+---------+
| OK     | NUCLEO_L433RC_P | ARM       | DTCT_1      | Simple detect test                    |     0.51     |    10   |
| OK     | NUCLEO_L433RC_P | ARM       | EXAMPLE_1   | /dev/null                             |     3.41     |    20   |
| OK     | NUCLEO_L433RC_P | ARM       | MBED_10     | Hello World                           |     0.35     |    5    |
| OK     | NUCLEO_L433RC_P | ARM       | MBED_11     | Ticker Int                            |    11.36     |    15   |
| OK     | NUCLEO_L433RC_P | ARM       | MBED_12     | C++                                   |     1.34     |    10   |
| OK     | NUCLEO_L433RC_P | ARM       | MBED_16     | RTC                                   |     4.59     |    20   |
| OK     | NUCLEO_L433RC_P | ARM       | MBED_2      | stdio                                 |     0.77     |    20   |
| OK     | NUCLEO_L433RC_P | ARM       | MBED_23     | Ticker Int us                         |    11.34     |    15   |
| OK     | NUCLEO_L433RC_P | ARM       | MBED_24     | Timeout Int us                        |    11.49     |    15   |
| OK     | NUCLEO_L433RC_P | ARM       | MBED_25     | Time us                               |    11.32     |    15   |
| OK     | NUCLEO_L433RC_P | ARM       | MBED_26     | Integer constant division             |     1.34     |    20   |
| OK     | NUCLEO_L433RC_P | ARM       | MBED_34     | Ticker Two callbacks                  |    11.36     |    15   |
| OK     | NUCLEO_L433RC_P | ARM       | MBED_37     | Serial NC RX                          |     6.91     |    20   |
| OK     | NUCLEO_L433RC_P | ARM       | MBED_38     | Serial NC TX                          |     5.88     |    20   |
| OK     | NUCLEO_L433RC_P | ARM       | MBED_A1     | Basic                                 |     1.36     |    20   |
| OK     | NUCLEO_L433RC_P | ARM       | MBED_A21    | Call function before main (mbed_main) |     1.41     |    20   |
| OK     | NUCLEO_L433RC_P | ARM       | MBED_A9     | Serial Echo at 115200                 |     1.72     |    20   |
| OK     | NUCLEO_L433RC_P | ARM       | MBED_BUSOUT | BusOut                                |     2.27     |    30   |
+--------+-----------------+-----------+-------------+---------------------------------------+--------------+---------+
+--------+-----------------+-----------+-------------+---------------------------------------+--------------+---------+
| Result | Target          | Toolchain | Test ID     | Test Description                      | Elapsed Time | Timeout |
+--------+-----------------+-----------+-------------+---------------------------------------+--------------+---------+
| OK     | NUCLEO_L433RC_P | GCC_ARM   | DTCT_1      | Simple detect test                    |     0.49     |    10   |
| OK     | NUCLEO_L433RC_P | GCC_ARM   | EXAMPLE_1   | /dev/null                             |     3.44     |    20   |
| OK     | NUCLEO_L433RC_P | GCC_ARM   | MBED_10     | Hello World                           |     0.34     |    5    |
| OK     | NUCLEO_L433RC_P | GCC_ARM   | MBED_11     | Ticker Int                            |    11.34     |    15   |
| OK     | NUCLEO_L433RC_P | GCC_ARM   | MBED_12     | C++                                   |     1.35     |    10   |
| OK     | NUCLEO_L433RC_P | GCC_ARM   | MBED_16     | RTC                                   |     4.61     |    20   |
| OK     | NUCLEO_L433RC_P | GCC_ARM   | MBED_2      | stdio                                 |     0.75     |    20   |
| OK     | NUCLEO_L433RC_P | GCC_ARM   | MBED_23     | Ticker Int us                         |    11.35     |    15   |
| OK     | NUCLEO_L433RC_P | GCC_ARM   | MBED_24     | Timeout Int us                        |    11.57     |    15   |
| OK     | NUCLEO_L433RC_P | GCC_ARM   | MBED_25     | Time us                               |    11.37     |    15   |
| OK     | NUCLEO_L433RC_P | GCC_ARM   | MBED_26     | Integer constant division             |     1.37     |    20   |
| OK     | NUCLEO_L433RC_P | GCC_ARM   | MBED_34     | Ticker Two callbacks                  |    11.38     |    15   |
| OK     | NUCLEO_L433RC_P | GCC_ARM   | MBED_37     | Serial NC RX                          |     6.92     |    20   |
| OK     | NUCLEO_L433RC_P | GCC_ARM   | MBED_38     | Serial NC TX                          |     5.89     |    20   |
| OK     | NUCLEO_L433RC_P | GCC_ARM   | MBED_A1     | Basic                                 |     1.33     |    20   |
| OK     | NUCLEO_L433RC_P | GCC_ARM   | MBED_A21    | Call function before main (mbed_main) |     1.42     |    20   |
| OK     | NUCLEO_L433RC_P | GCC_ARM   | MBED_A9     | Serial Echo at 115200                 |     1.7      |    20   |
| OK     | NUCLEO_L433RC_P | GCC_ARM   | MBED_BUSOUT | BusOut                                |     2.28     |    30   |
+--------+-----------------+-----------+-------------+---------------------------------------+--------------+---------+
+--------+-----------------+-----------+-------------+---------------------------------------+--------------+---------+
| Result | Target          | Toolchain | Test ID     | Test Description                      | Elapsed Time | Timeout |
+--------+-----------------+-----------+-------------+---------------------------------------+--------------+---------+
| OK     | NUCLEO_L433RC_P | IAR       | DTCT_1      | Simple detect test                    |     0.48     |    10   |
| OK     | NUCLEO_L433RC_P | IAR       | EXAMPLE_1   | /dev/null                             |     3.41     |    20   |
| OK     | NUCLEO_L433RC_P | IAR       | MBED_10     | Hello World                           |     0.33     |    5    |
| OK     | NUCLEO_L433RC_P | IAR       | MBED_11     | Ticker Int                            |    11.37     |    15   |
| OK     | NUCLEO_L433RC_P | IAR       | MBED_12     | C++                                   |     1.33     |    10   |
| OK     | NUCLEO_L433RC_P | IAR       | MBED_16     | RTC                                   |     4.61     |    20   |
| OK     | NUCLEO_L433RC_P | IAR       | MBED_2      | stdio                                 |     0.72     |    20   |
| OK     | NUCLEO_L433RC_P | IAR       | MBED_23     | Ticker Int us                         |    11.33     |    15   |
| OK     | NUCLEO_L433RC_P | IAR       | MBED_24     | Timeout Int us                        |    11.52     |    15   |
| OK     | NUCLEO_L433RC_P | IAR       | MBED_25     | Time us                               |    11.36     |    15   |
| OK     | NUCLEO_L433RC_P | IAR       | MBED_26     | Integer constant division             |     1.33     |    20   |
| OK     | NUCLEO_L433RC_P | IAR       | MBED_34     | Ticker Two callbacks                  |    11.37     |    15   |
| OK     | NUCLEO_L433RC_P | IAR       | MBED_37     | Serial NC RX                          |     6.94     |    20   |
| OK     | NUCLEO_L433RC_P | IAR       | MBED_38     | Serial NC TX                          |     5.9      |    20   |
| OK     | NUCLEO_L433RC_P | IAR       | MBED_A1     | Basic                                 |     1.33     |    20   |
| OK     | NUCLEO_L433RC_P | IAR       | MBED_A21    | Call function before main (mbed_main) |     1.41     |    20   |
| OK     | NUCLEO_L433RC_P | IAR       | MBED_A9     | Serial Echo at 115200                 |     1.69     |    20   |
| OK     | NUCLEO_L433RC_P | IAR       | MBED_BUSOUT | BusOut                                |     2.28     |    30   |
+--------+-----------------+-----------+-------------+---------------------------------------+--------------+---------+
```

mbedOS 5 tests are coming...

helmut64 commented:
@LMESTM I added the boot loader support into the 433RC_P and 432KC L4 targets. In terms of memory and flash sizes there is no difference between the two MCUs therefore I added bootloader support for these targets. Now we support identical boot loader features for all L4 targets.
I tested this version using the GCC environment and it works fine.

LMESTM commented:
@helmut64 the change looks good to me, thanks !

bcostm commented:
mbedOS 5 tests are OK. There is only one problem with tests-mbed_drivers-lp_timeout test (1sec LowPowerTimeout from deepsleep). To be corrected as this test is OK with all L4 devices (results of tests performed on mbed-os 5.5.6).

[report_test5_all_NUCLEO_L433RC_P_ARM.txt](https://github.com/ARMmbed/mbed-os/files/1359219/report_test5_all_NUCLEO_L433RC_P_ARM.txt)
[report_test5_all_NUCLEO_L433RC_P_GCC_ARM.txt](https://github.com/ARMmbed/mbed-os/files/1359220/report_test5_all_NUCLEO_L433RC_P_GCC_ARM.txt)
[report_test5_all_NUCLEO_L433RC_P_IAR.txt](https://github.com/ARMmbed/mbed-os/files/1359332/report_test5_all_NUCLEO_L433RC_P_IAR.txt)

helmut64 commented:
@bcostm Thank you for all the help so far, it looks already very promising.
I found the test code and will debug it. For deepsleep I have some experience. I have not used the LowPowerTimer so far, but will look into it.

helmut64 commented:
In my App I still have an issue when using deepsleep (433RC_P) with an error coming from Ticker->detach:
"Deep sleep lock would underflow (< 0)"
If I use only sleep it does not occur.

This problem does not occur on target NUCLEO_L476RG. I will verify it Monday with the 432KC target, to check what is up with the 432KC.

Maybe it is related to the mbed os 5 test results. I investigate further into it.

bcostm commented:
Yes it seems related but I don't know why ?
I've just ran the mbedOS 5 lp-timeout tests on the NUCLEO_L432KC and they are OK. Tested on mbed master branch:
```
+-----------------------+---------------+-------------------------------+-------------------------------------+--------+--------+--------+--------------------+
| target                | platform_name | test suite                    | test case                           | passed | failed | result | elapsed_time (sec) |
+-----------------------+---------------+-------------------------------+-------------------------------------+--------+--------+--------+--------------------+
| NUCLEO_L432KC-GCC_ARM | NUCLEO_L432KC | tests-mbed_drivers-lp_timeout | 1ms LowPowerTimeout                 | 1      | 0      | OK     | 0.05               |
| NUCLEO_L432KC-GCC_ARM | NUCLEO_L432KC | tests-mbed_drivers-lp_timeout | 1sec LowPowerTimeout                | 1      | 0      | OK     | 1.04               |
| NUCLEO_L432KC-GCC_ARM | NUCLEO_L432KC | tests-mbed_drivers-lp_timeout | 1sec LowPowerTimeout from deepsleep | 1      | 0      | OK     | 1.07               |
| NUCLEO_L432KC-GCC_ARM | NUCLEO_L432KC | tests-mbed_drivers-lp_timeout | 1sec LowPowerTimeout from sleep     | 1      | 0      | OK     | 1.06               |
| NUCLEO_L432KC-GCC_ARM | NUCLEO_L432KC | tests-mbed_drivers-lp_timeout | 500us LowPowerTimeout               | 1      | 0      | OK     | 0.04               |
| NUCLEO_L432KC-GCC_ARM | NUCLEO_L432KC | tests-mbed_drivers-lp_timeout | 5sec LowPowerTimeout                | 1      | 0      | OK     | 5.06               |
+-----------------------+---------------+-------------------------------+-------------------------------------+--------+--------+--------+--------------------+
```

helmut64 commented:
The "Deep sleep lock would underflow (< 0)" problem was a home made problem only in my dev environment, including wrong header files in my project (from the mbed 151 instead of 433_P head). 
I also fixed a missing ifdef USART3_BASE case for the 432.

Finally I will resume investigation into the tests-mbed_drivers-lp_timeout problem.

helmut64 commented:
I tested the deepsleep and lp_timer features and cannot find any problems, here is my mbed test code:
```
#include "mbed.h"

static volatile bool complete;
static volatile timestamp_t complete_time;
static ticker_event_t delay_event;
static const ticker_data_t *lp_ticker_data = get_lp_ticker_data();
static LowPowerTimer lp_timer;

#define LONG_TIMEOUT (100000)


void cb_done_deepsleep(uint32_t id) {
    if ((uint32_t)&delay_event == id) {
        complete_time = lp_timer.read_us();
        complete = true;
    } else {
        // Normal ticker handling
        TimerEvent::irq(id);
    }
}

void lp_ticker_1s_deepsleep()
{
    Serial ser(USBTX, USBRX);
    ser.baud(115200*2);
    
    complete = false;
    uint32_t delay_ts;

#if 0
    lp_timer.reset();
    lp_timer.start();
    ser.printf("lp_timer 0: %d\r\n", lp_timer.read_us());
    wait_ms(1000);
    ser.printf("lp_timer 1: %d\r\n", lp_timer.read_us());
    wait_ms(1000);
    ser.printf("lp_timer 2: %d\r\n", lp_timer.read_us());
    wait_ms(1000);
    ser.printf("lp_timer 3: %d\r\n", lp_timer.read_us());
    wait_ms(1000);
    ser.printf("lp_timer 4: %d\r\n", lp_timer.read_us());
#endif
    /*
     * Since deepsleep() may shut down the UART peripheral, we wait for 10ms
     * to allow for hardware serial buffers to completely flush.
     
     * This should be replaced with a better function that checks if the
     * hardware buffers are empty. However, such an API does not exist now,
     * so we'll use the wait_ms() function for now.
     */
    wait_ms(10);
    
    ticker_set_handler(lp_ticker_data, cb_done_deepsleep);
    ticker_remove_event(lp_ticker_data, &delay_event);
    delay_ts = ticker_read(lp_ticker_data) + 1000000;
    
    /*
     * We use here the low power timer instead of microsecond timer for start and
     * end because the microseconds timer might be disable during deepsleep.
     */
    lp_timer.reset();
    lp_timer.start();
    ticker_insert_event(lp_ticker_data, &delay_event, delay_ts, (uint32_t)&delay_event);
    deepsleep();
    while (!complete);
    lp_timer.stop();
    
    //TEST_ASSERT_UINT32_WITHIN(LONG_TIMEOUT, 1000000, complete_time);
    //TEST_ASSERT_TRUE(complete);
    ser.printf("\b\b\b\b\b complete_time: %d\r\n         ", (int)complete_time);
}

int main(void)
{
    for (;;) {
    	lp_ticker_1s_deepsleep();
    }
}
```

helmut64 commented:
PS: The printing \b and space are because deepsleep turns of the ser->printf UART and produces garbage chars when turning on or off, this eliminates it.

bcostm commented:
I tested your code and it looks ok for me too. I have also added a print of the timer read value in the mbed_drivers-lp_timeout test and the value is the same as in your code and ok. So, there is something else which causes this test to be fail but I don't know what ?

0xc0170 commented:
@helmut64 I am trying to understand what is outstanding for this new target. After reading the latest comments, seems all issues are resolved, or something still there is to fix (known issue) ?

bcostm commented:
Another thing: I have executed the `tests-mbed_drivers-lp_timeout` on the NUCLEO_L432KC using this branch and the test is FAIL too.

I have then re-checked the same test on the `master` branch and it is PASS.

So, there is something in this branch that causes the tests-mbed_drivers-lp_timeout to be FAIL on both NUCLEO_L433RC_P and NUCLEO_L432KC boards.

helmut64 commented:
@bcostm @0xc0170 My proposal is the following:
Lets declare this branch for the new NUCLEO_L433RC-P final and complete and publish the mbed support for NUCLEO_L433RC-P board.
I am happy to continue to do support and my projects will use this new target in real word daliy work using the offline compiler, Keil MDK and GCC build environments.
For additional or new issues we just open a new case.

I am very happy with the new support. Thanks for everyone helping.

PS: This new platform support will also complete the issue #450 (STM32L4 combining support for L431xx, L432xx and L433xx devices)

helmut64 commented:
@bcostm the only changes for the for the L432 in this branch are:
- enabled SERIAL_ASYNCH
- bootloader support added
I added this because it has been also added in the L433 and the MCUs are identical for this kind of features.

helmut64 commented:
Another note: Once this target is published I will switch my projects from the WakeUpRTC timer to the LowPowerTimer (lp_timer) to get this active in my project.

studavekar commented:
/morph test

0xc0170 commented:
Failure not related to this patch. I would restart the test. But not certain about this comment:

> Lets declare this branch for the new NUCLEO_L433RC-P final and complete and publish the mbed support for the NUCLEO_L433RC-P board.

Should this be send to another branch, but master?

helmut64 commented:
@0xc0170 Martin, bcostm completed test under mbed os 5 reports a problem that the low power ticker does not wakup on time, I believe the wakeup may be too late. Usually it should wakeup after one second. I did extra testing and for me it works always fine within the one second range.

I just like to finish this board support because it works great and I cannot find any problems. In case a reproducible problem occurs I am happy to investigate into it.

helmut64 commented:
It should be merged into master.

0xc0170 commented:
/morph build

0xc0170 commented:
/morph uvisor-test

0xc0170 commented:
@bcostm Are you guys happy with this PR?

bcostm commented:
@helmut64 Can you please rebase on top of master ? Due to recent changes in the NUCLEO_L432KC platform that has been merged in master and that your PR overwrites. Thanks.

helmut64 commented:
@bcostm I may need some help regarding the rebase which I have not done before.

On github.com I forked the mbed-os repro which is now on Github helmut64/mbed-os
This I pulled to my local development desktop, created the NUCLEO_STM32L433_SUPPORT branch pushed by local changes patches back to Github helmut64/mbed-os

How do I rebase the Github helmut64/mbed-os fork?

Thank you Helmut

LMESTM commented:
@helmut64 
Here is what I'd do: 
In the same mbed repo from local development desktop, add also the mbed-os reference:
`` git remote add mbed https://github.com/ARMmbed/mbed-os ``
To make sure it worked ok, check that fetch works properly
`` git fetch mbed``
If this ok, create a backup of your current branch (NUCLEO_STM32L433_SUPPORT) by creating a new one 
at the same position (no really needed as you have pushed it but well ... )
`` git checkout -b stm32l433_backup ``
Then start rebase 
`` git pull --rebase mbed master``
This is where fun starts ... you'll maybe have to solve conflicts (I use git merge-tool command) and after solving each conflict, you'll need the command below
`` git rebase --continue``
until rebase is complete. 
If this completes well, you can push again to github ...

theotherjimmy commented:
labeled "needs: work" as we are waiting on the rebase.

helmut64 commented:
I am working on it, however this week I am busy in the US

helmut64 commented:
I did the rebase, it worked without any merge conflicts. I hope that I did the rebase correct.

0xc0170 commented:
> I did the rebase, it worked without any merge conflicts. I hope that I did the rebase correct.

Please review the commits, github shows that here are commits that are already on master. Seems like the rebase was not done correctly.

theotherjimmy commented:
@helmut64 Would you like some assistance on that rebase?

helmut64 commented:
@theotherjimmy Dear Jim, yes help would be pretty much appreciated. Maybe I explain it a little bit.

The regular procedure is the following. On github mbed-os I forked a copy into my helmut64 github repository which ends up in helmut64/mbed-os. The next step is that I cloned helmut64/mbed-os to my local Mac. There I created the branch, did the changes and pushed this back.

My first question is, once I have forked from ARMmbed/mbed-os to helmut64/mbed-os, how can I update helmut64/mbed-os on github to get this to the latest ARMmbed/mbed-os version. This is one of my problems.

In the past I removed helmut64/mbed-os and started all over again, however that’s not good.

I did a local tar backup of the local github repository before I did the rebase.
Regards Helmut

helmut64 commented:
One option is that I start over again and redo the changes, because I have it all documented.

bcostm commented:
What I do.

1) my remotes are:
```
$ git remote -v
ARMmbed https://github.com/ARMmbed/mbed-os.git (fetch)
ARMmbed https://github.com/ARMmbed/mbed-os.git (push)
bcostm  https://github.com/bcostm/mbed-os.git (fetch)
bcostm  https://github.com/bcostm/mbed-os.git (push)
```
Instead of bcostm you should have helmut64.

2) Then
`$ git pull --rebase ARMmbed master`

Sometimes I do : `$ git rebase -i ARMmbed/master` to remove/change some commits

3) I solve all conflicts if any

4) Then `git push` or `git push -f`

theotherjimmy commented:
@bcostm Thanks for responding. @helmut64 I recommend following @bcostm's comment above.

helmut64 commented:
I will work on this, thank you for the advise.

helmut64 commented:
I did the rebase. I hope that this is now good.
I compiled the branch, it contains my changes.

0xc0170 commented:
@helmut64 The branch shows here 250+ commits, it should contain only your changes.

helmut64 commented:
@0xc0170 Hello Martin, than I will create a new branch, redo the changes and push it.
Will this be working for you?

0xc0170 commented:
> @0xc0170 Hello Martin, than I will create a new branch, redo the changes and push it.
Will this be working for you?

Whatever works for you. If a new branch, then this would be closed and new PR opened.

helmut64 commented:
I created a new pull request #5591 branch NUCLEO_STM32L433_NEW
